Userland/Libraries/LibCore/FileSystem.cpp

```diff
@@ -0,0 +1,449 @@
+        return false;
+    auto st = st_or_error.release_value();
+    return S_ISBLK(st.st_mode) || S_ISCHR(st.st_mode);
+}
+
+bool is_block_device(StringView filename)
+{
+    auto st_or_error = System::stat(filename);
+    if (st_or_error.is_error())
+        return false;
+    auto st = st_or_error.release_value();
+    return S_ISBLK(st.st_mode);
+}
+
+bool is_block_device(int fd)
+{
+    auto st_or_error = System::fstat(fd);
+    if (st_or_error.is_error())
+        return false;
+    auto st = st_or_error.release_value();
+    return S_ISBLK(st.st_mode);
+}
+
+bool is_char_device(StringView filename)
+{
+    auto st_or_error = System::stat(filename);
+    if (st_or_error.is_error())
+        return false;
+    auto st = st_or_error.release_value();
+    return S_ISCHR(st.st_mode);
+}
+
+bool is_char_device(int fd)
+{
+    auto st_or_error = System::fstat(fd);
+    if (st_or_error.is_error())
+        return false;
+    auto st = st_or_error.release_value();
+    return S_ISCHR(st.st_mode);
+}
+
+bool is_directory(StringView filename)
+{
+    auto st_or_error = System::stat(filename);
+    if (st_or_error.is_error())
+        return false;
+    auto st = st_or_error.release_value();
+    return S_ISDIR(st.st_mode);
+}
+
+bool is_directory(int fd)
+{
+    auto st_or_error = System::fstat(fd);
+    if (st_or_error.is_error())
+        return false;
+    auto st = st_or_error.release_value();
+    return S_ISDIR(st.st_mode);
+}
+
+bool is_link(StringView filename)
+{
+    auto st_or_error = System::lstat(filename);
+    if (st_or_error.is_error())
+        return false;
+    auto st = st_or_error.release_value();
+    return S_ISLNK(st.st_mode);
+}
+
+bool is_link(int fd)
+{
+    struct stat st;
+    if (fstat(fd, &st) < 0)
+        return false;
+    return S_ISLNK(st.st_mode);
+}
+
+static DeprecatedString get_duplicate_name(DeprecatedString const& path, int duplicate_count)
+{
+    if (duplicate_count == 0) {
+        return path;
+    }
+    LexicalPath lexical_path(path);
+    StringBuilder duplicated_name;
+    duplicated_name.append('/');
+    auto& parts = lexical_path.parts_view();
+    for (size_t i = 0; i < parts.size() - 1; ++i) {
+        duplicated_name.appendff("{}/", parts[i]);
+    }
+    auto prev_duplicate_tag = DeprecatedString::formatted("({})", duplicate_count);
+    auto title = lexical_path.title();
+    if (title.ends_with(prev_duplicate_tag)) {
+        // remove the previous duplicate tag "(n)" so we can add a new tag.
+        title = title.substring_view(0, title.length() - prev_duplicate_tag.length());
+    }
+    duplicated_name.appendff("{} ({})", title, duplicate_count);
+    if (!lexical_path.extension().is_empty()) {
+        duplicated_name.appendff(".{}", lexical_path.extension());
+    }
+    return duplicated_name.to_deprecated_string();
+}
+
+ErrorOr<void, CopyError> copy_file(StringView dst_path, StringView src_path, struct stat const& src_stat, NonnullOwnPtr<File>& source, PreserveMode preserve_mode)
```

Comment:
Taking a reference to a NonnullOwnPtr is not a good idea, I think you can take a regular reference instead which NonnullOwnPtr can provide without problems.